code/modules/antagonists/wraith/objs/machinery/vortex_wraith.dm

```diff
@@ -141,36 +138,223 @@
 	disposing()
 		if (src.master != null)
 			src.master.linked_portal = null
+			var/datum/targetable/ability = src.master.abilityHolder.getAbility(/datum/targetable/wraithAbility/create_summon_portal)
+			ability.doCooldown()
 		deleteLinkedCritters()
+		STOP_TRACKING
 		. = ..()
 
-	proc/getMobValue(var/obj/O)
-		switch (O)
-			if (/obj/critter/bear)
-				return 10
-			if (/obj/critter/wraithskeleton)
-				return 4
-			if (/obj/critter/shade)
-				return 4
-			if (/obj/critter/crunched)
-				return 4
-			if (/obj/critter/bat/buff)
-				return 3
-			if (/obj/critter/lion)
-				return 5
-			if (/obj/critter/brullbar)
-				return 15
-			if (/obj/critter/gunbot/heavy)
-				return 15
-			if (/obj/critter/ancient_thing)
-				return 7
-			if (/obj/critter/mechmonstrositycrawler)
-				return 4
-			else	//You never know, lets give an average point cost
-				return 6
+	attack_hand(mob/user)
+		return
 
 	proc/deleteLinkedCritters()
 		for (var/obj/critter/C in src.critter_list)
 			animate(C, alpha=0, time=2 SECONDS)
 			SPAWN(2 SECOND)
 				qdel(C)
+		for (var/obj/critter/C in src.strong_critter_list)
+			animate(C, alpha=0, time=2 SECONDS)
+			SPAWN(2 SECOND)
+				qdel(C)
+
+/obj/machinery/wraith/vortex_wraith/ui_status(mob/user, datum/ui_state/state)
+	if(IN_RANGE(user, src, 8))
+		. = tgui_broken_state.can_use_topic(src, user)
+
+/obj/machinery/wraith/vortex_wraith/ui_data(mob/user)
+	var/cooldown = FALSE
+	if (src.last_upgrade > TIME)
+		cooldown = TRUE
+	. = list(
+		"spawnrate" = (src.spawn_rate / 10),
+		"spawnrange" = src.spawn_radius,
+		"mob_value_cap" = src.mob_value_cap,
+		"_health" = src._health,
+		"maxhealth" = src.max_health,
+		"summon_power" = src.summon_power,
+		"upgrade_cost" = src.upgrade_cost,
+		"active" = src.active,
+		"cooldown" = cooldown
+	)
+
+/obj/machinery/wraith/vortex_wraith/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "VortexWraith")
+		ui.open()
+
+/obj/machinery/wraith/vortex_wraith/ui_act(action, params)
+	. = ..()
+	if (.)
+		return
+	if (!istype(usr, /mob/living/intangible/wraith))
+		boutput(usr, "<span class='alert'>You shouldn't be seeing this! Dial 1-800 coder.</span>")
+		. = TRUE
+		return
+	else
+		var/datum/abilityHolder/wraith/AH = usr.abilityHolder
+		switch(action)
+			if ("up_spawnrate")
+				if (TIME < src.last_upgrade)
+					boutput(usr, "<span class='alert'>You just upgraded your portal. Give it a minute.</span>")
+					. = TRUE
+					return
+				else if (AH.points < src.upgrade_cost)
+					boutput(usr, "<span class='alert'>You do not have enough points to upgrade that!</span>")
+					. = TRUE
+					return
+				else
+					if (src.spawn_rate <= 25 SECONDS)
+						boutput(usr, "<span class='notice'>Maximum spawn rate already achieved!</span>")
+						. = TRUE
+						return
+					else
+						src.spawn_rate -= 3 SECONDS
+						src.last_upgrade = TIME + src.upgrade_cooldown
```

Comment:
relevant line to above comment (no changes needed here other than the var rename)